### PR TITLE
Standard deviation for aggregations

### PR DESCRIPTION
**Pull request by ashtul**
This PR is a copy of #125 which was opened on June 6th.

### CODE REVIEW

**Comment by ashtul** [comment]
Part of enhancement requested on issue  #84

**Comment by ashtul** [comment]
@danni-m 
Resolved conflict.

File: src/compaction.c [hunk 1]
```diff
@@ -26,9 +26,9 @@ typedef struct AvgContext {
 } AvgContext;
 
 typedef struct StdContext {
-    double std;
-    double mean;
-    double cnt;
+    double sum;
+    double sum_2;
```
Comment (otnielvh) [inline src/compaction.c]: sum_2 will be confusing for a future reader. Maybe add a comment to explain more

File: src/compaction.c [hunk 2]
```diff
@@ -98,34 +98,38 @@ void AvgReadContext(void *contextPtr, RedisModuleIO * io){
...
-    if(context->cnt > 0) {
-        res = context->std / context->cnt;
+    double sum = context->sum;
+    double sum_2 = context->sum_2;
+    uint64_t count = context->cnt;
+    if(count <= 1) {
+        return 0;
     }
-    return res;
+    return (sum_2 - 2 * (sum / count) * sum + pow(sum / count, 2) * count) / count;
```
Comment (otnielvh) [inline src/compaction.c]: You can add here a comment something like:
`var(X) = sum((x_i - E[X])^2) =  sum(x_i)^2 -2sum(x_i)*E[X] + E^2[X]`

**Comment by ashtul** [comment]
@danni-m,
@otnielvh have reviewed the code and asked for changes which were implemented.

File: src/compaction.c (hunk 1 above)
Comment (ashtul) [reply src/compaction.c]: comment added

File: src/compaction.c (hunk 2 above)
Comment (ashtul) [reply src/compaction.c]: Done!

**Comment by ashtul** [comment]
@danni-m @gkorland 
Can this branch be merged? It does not break API.

File: src/compaction.h [hunk 3]
```diff
@@ -16,9 +16,13 @@
 #define AGG_MAX 2
 #define AGG_SUM 3
 #define AGG_AVG 4
-#define AGG_COUNT 5
-#define AGG_FIRST 6
-#define AGG_LAST 7
+#define AGG_STD_P 5
```
Comment (danni-m) [inline src/compaction.h]: Why not add them in the end?

File: src/consts.h [hunk 4]
```diff
@@ -30,11 +30,15 @@ typedef enum {
...
+    TS_AGG_STD_P,
+    TS_AGG_STD_S,
+    TS_AGG_VAR_P,
+    TS_AGG_VAR_S,
     TS_AGG_COUNT,
     TS_AGG_FIRST,
     TS_AGG_LAST,
     TS_AGG_RANGE,
-    TS_AGG_TYPES_MAX // 9
+    TS_AGG_TYPES_MAX // 10
```
Comment (danni-m) [inline src/consts.h]: definitely not 10

File: src/compaction.h (hunk 3 above)
Comment (ashtul) [reply src/compaction.h]: My thought was, standard deviation belongs with average so I put them together.

File: src/consts.h (hunk 4 above)
Comment (ashtul) [reply src/consts.h]: You are right. I missed this when expanded with more aggs.

Comment (ashtul) [reply src/consts.h]: was corrected

File: src/tests/test_module.py [hunk 5]
```diff
@@ -537,6 +537,42 @@ def test_agg_avg(self):
             actual_result = r.execute_command('TS.RANGE', agg_key, 10, 50)
             assert expected_result == actual_result
 
+    def test_agg_std_p(self):
+        with self.redis() as r:
+            agg_key = self._insert_agg_data(r, 'tester', 'std.p')
```
Comment (otnielvh) [inline src/tests/test_module.py]: Too much hard coded stuff in the test, which will make it very flaky.
I recommend defining a list with values and inserting it into the redis, then computing the std using python standard library and the implementation you did and assert they have similar values.

File: src/tests/test_module.py [hunk 6]
```diff
@@ -537,6 +537,42 @@ def test_agg_avg(self):
...
+
+            expected_result = [[10, '25.869'], [20, '25.869'], [30, '25.869'], [40, '25.869']]
+            actual_result = r.execute_command('TS.RANGE', agg_key, 10, 50)
+            for i in range(len(expected_result)):
+                assert abs(float(expected_result[i][1]) - float(actual_result[i][1])) < 0.001
```
Comment (otnielvh) [inline src/tests/test_module.py]: defined `EPSILON = 0.001`, please don't use magic numbers here

File: docs/index.md [hunk 7]
```diff
@@ -7,7 +7,7 @@ RedisTimeSeries is a Redis Module adding a Time Series data structure to Redis.
 Read more about the v1.0 GA features [here](https://redislabs.com/blog/redistimeseries-ga-making-4th-dimension-truly-immersive/).
 - High volume inserts, low latency reads
 - Query by start time and end-time
-- Aggregated queries (Min, Max, Avg, Sum, Range, Count, First, Last) for any time bucket
+- Aggregated queries (Min, Max, Avg, Sum, Range, Count, First, Last, Std.P, Std.S, Var.P, Var.S) for any time bucket
```
[the remark this answers is not in the document]
Comment (ashtul) [reply docs/index.md]: agree

File: docs/commands.md [hunk 8]
```diff
@@ -159,12 +159,12 @@ If this command is used to add data to an existing timeseries, `retentionTime` a
...
 ```sql
-TS.CREATERULE sourceKey destKey AGGREGATION aggType timeBucket
+TS.CREATERULE sourceKey destKey AGGREGATION aggregationType timeBucket
 ```
 
 - sourceKey - Key name for source time series
 - destKey - Key name for destination time series
-- aggType - Aggregation type: avg, sum, min, max, range, count, first, last
+- aggregationType - Aggregation type: avg, sum, min, max, range, count, first, last, std.p, std.s, var.p, var.s
 - timeBucket - Time bucket for aggregation in milliseconds
```
[the remark this answers is not in the document]
Comment (ashtul) [reply docs/commands.md]: They offer both options on other timeseries DBs

File: src/compaction.c [hunk 9]
```diff
@@ -76,19 +83,87 @@ void AvgReset(void *contextPtr) {
...
+    return variance(context->sum, context->sum_2, count) / count;
+}
+
+double VarSamplesFinalize(void *contextPtr) {
+    StdContext *context = (StdContext *)contextPtr;
+    uint64_t count = context->cnt;
+    if(count <= 1) {
+        return 0;
+    }
+    return variance(context->sum, context->sum_2, count) / (count - 1);
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/compaction.c]: Nice idea

File: src/compaction.c [hunk 10]
```diff
@@ -76,19 +83,87 @@ void AvgReset(void *contextPtr) {
...
+    StdContext *context = (StdContext *)contextPtr;
+    ++context->cnt;
+    context->sum += value;
+    context->sum_2 += value * value; 
+}
+
+static double variance(double sum, double sum_2, double count) {
+    /*  var(X) = sum((x_i - E[X])^2)
+     *  = sum(x_i^2) - 2 * sum(x_i) * E[X] + E^2[X] */
+    return  sum_2 - 2 * sum * (sum / count) + pow(sum / count, 2) * count;
```
[the remark this answers is not in the document]
Comment (otnielvh) [reply src/compaction.c]: @ashtul You're right about the first one, I forgot the sum....

File: src/compaction.c [hunk 11]
```diff
@@ -76,19 +83,87 @@ void AvgReset(void *contextPtr) {
...
 
-void AvgReadContext(void *contextPtr, RedisModuleIO * io){
+void AvgReadContext(void *contextPtr, RedisModuleIO *io){
     AvgContext *context = (AvgContext *)contextPtr;
     context->val = RedisModule_LoadDouble(io);
     context->cnt = RedisModule_LoadDouble(io);
 }
 
-void rm_free(void* ptr) {
+void *StdCreateContext() {
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/compaction.c]: Kept naming of other aggregators

File: src/compaction.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (ashtul) [reply src/compaction.c]: Changed accordingly

File: src/tests/test_module.py (hunk 6 above)
Comment (ashtul) [reply src/tests/test_module.py]: Defined a constant

File: src/compaction.c [hunk 12]
```diff
@@ -122,7 +122,7 @@ double VarPopulationFinalize(void *contextPtr) {
     if(count <= 1) {
         return 0;
     }
```
Comment (otnielvh) [inline src/compaction.c]: This test should be handled by `variance` function (before dividing by `count`)

File: src/compaction.c [hunk 13]
```diff
@@ -131,7 +131,7 @@ double VarSamplesFinalize(void *contextPtr) {
...
```
Comment (otnielvh) [inline src/compaction.c]: same as previous comment

**Comment by otnielvh** [comment]
Looks good, just one minor comment

File: src/compaction.c [hunk 14]
```diff
@@ -76,19 +83,87 @@ void AvgReset(void *contextPtr) {
...
+    return  sum_2 - 2 * sum * (sum / count) + pow(sum / count, 2) * count;
+}
+
+double VarPopulationFinalize(void *contextPtr) {
+    StdContext *context = (StdContext *)contextPtr;
+    uint64_t count = context->cnt;
+    if(count <= 1) {
+        return 0;
+    }
+    return variance(context->sum, context->sum_2, count) / count;
```
[the remark this answers is not in the document]
Comment (ashtul) [reply src/compaction.c]: Done

File: src/compaction.c (hunk 12 above)
Comment (ashtul) [reply src/compaction.c]: Nice catch!
Done.

**Comment by ashtul** [comment]
@otnielvh, as requested, made a random test with seed and epsilon.

File: src/tests/test_module.py [hunk 15]
```diff
@@ -537,6 +541,66 @@ def test_agg_avg(self):
...
+        
+            stdev = statistics.stdev(items)
+            var = statistics.variance(items)
+            assert r.execute_command('TS.CREATE', raw_key)
+            assert r.execute_command('TS.CREATE', std_key)
+            assert r.execute_command('TS.CREATE', var_key)
+            assert r.execute_command('TS.CREATERULE', raw_key, std_key, "AGGREGATION", 'std.s', random_numbers)
+            assert r.execute_command('TS.CREATERULE', raw_key, var_key, "AGGREGATION", 'var.s', random_numbers)
+    
+            for i in range(random_numbers):
```
Comment (otnielvh) [inline src/tests/test_module.py]: `len(items)` would be clearer

**Pull request merged**